### PR TITLE
Add Fluid Capability

### PR DESCRIPTION
**Pull request by mezz**
Builds on the previous PR #2459 and fixes some bugs.
Adds wrappers to turn every kind of legacy fluid handling into the new `IFluidHandler`.
Adds methods in `FluidUtils` to handle common cases,

I made a branch of Forestry to test out the legacy handling and new features.
https://github.com/ForestryMC/ForestryMC/tree/fluid-capability
This helped me define better methods in Forge's `FluidUtils` to handle some more complicated cases cleanly.
If you test with Forestry, note that its capsules, cans, etc are consumable and single use. This is intentional. It was not possible before this PR without special handling from the machines using them.

~~I have not used `IFluidHandler.getTankInfo()` at all. It seems sort of useful, but in practice I'm not sure if it's needed or not. Maybe pipes have a need for it, they're the only thing I haven't really tried out.~~ Improved the tank info after talking with people.

I @mezz consent to the licensing of my contributions to the Forge project under the license as stated in LICENSE-new.txt

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -27,7 +27,7 @@ apply plugin: "net.minecraftforge.gradle.launch4j"
 
 minecraft {
     version = "1.9.4"
-    mappings = 'snapshot_nodoc_20160518'
+    mappings = 'snapshot_nodoc_20160525'
```
Comment (asiekierka) [inline build.gradle]: That should probably not be here

File: patches/minecraft/net/minecraft/item/ItemStack.java.patch [hunk 2]
```diff
@@ -239,12 +239,12 @@
          }
      }
 +
-+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, net.minecraft.util.EnumFacing facing)
++    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
```
Comment (asiekierka) [inline patches/minecraft/net/minecraft/item/ItemStack.java.patch]: All those @Nullable probably shouldn't be in this PR

File: src/main/java/net/minecraftforge/fluids/FluidTank.java [hunk 3]
```diff
@@ -175,11 +198,28 @@ public FluidStack drain(int maxDrain, boolean doDrain)
...
+
             if (tile != null)
             {
                 FluidEvent.fireEvent(new FluidEvent.FluidDrainingEvent(fluid, tile.getWorld(), tile.getPos(), this, drained));
             }
         }
         return stack;
     }
+
+    public boolean canFill(Fluid fluid)
```
Comment (asiekierka) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: Do we need canFill, when fill has a simulation parameter?

File: src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java [hunk 4]
```diff
@@ -3,10 +3,13 @@
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.fluids.capability.templates.FluidTankItemStack;
 
 /**
  * Reference implementation of {@link IFluidContainerItem}. Use/extend this or implement your own.
+ * @deprecated See {@link net.minecraftforge.fluids.capability.ItemFluidContainer}
  */
+@Deprecated
```
Comment (asiekierka) [inline src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java]: I'm pretty sure you could mkae ItemFluidContainer provide a wrapping capability automatically.

File: src/main/java/net/minecraftforge/fluids/capability/ItemFluidContainer.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package net.minecraftforge.fluids.capability;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.common.capabilities.ICapabilityProvider;
+import net.minecraftforge.fluids.capability.templates.FluidTankItemStack;
+
+public class ItemFluidContainer extends Item
```
Comment (asiekierka) [inline src/main/java/net/minecraftforge/fluids/capability/ItemFluidContainer.java]: I don't think we need this class in Forge itself.

File: src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerConcatenate.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        return tanks.toArray(new FluidTankInfo[tanks.size()]);
+    }
+
+    @Override
+    public int fill(FluidStack resource, boolean doFill)
+    {
+        int k = 0;
+        for (IFluidHandler handler : subHandlers)
+        {
+            k += handler.fill(resource, doFill);
```
Comment (asiekierka) [inline src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerConcatenate.java]: Won't this break horribly and cause dupe bugs as k is never subtracted from the resource?

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java [hunk 7]
```diff
@@ -249,7 +249,7 @@ public static int getSkyBlendColour(World world, BlockPos center)
             for (int z = -distance; z <= distance; ++z)
             {
                 BlockPos pos = center.add(x, 0, z);
-                Biome biome = world.getBiomeGenForCoords(pos);
+                Biome biome = world.getBiome(pos);
```
Comment (gigaherz) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: This change looks unrelated to fluids? Seems like it belongs on a separate PR ;P

**Comment by williewillus** [comment]
Does this resolve the item issues brought up by Lex in team's PR? That was what was preventing it's pull.

**Comment by Actuarius** [comment]
@williewillus added labels [Feature, Fluids]

File: src/main/java/net/minecraftforge/fluids/FluidTank.java (hunk 3 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: canFill/canDrain are just for filtering basically. They defines the types of fluids that the tanks allows to flow in and out.

File: src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java (hunk 4 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java]: Good idea.

File: src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerConcatenate.java (hunk 6 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/capability/templates/FluidHandlerConcatenate.java]: Yes, this looks like a bug.

**Comment by mezz** [comment]
I believe this resolves the issues brought up in the other PR, yes.
The wrappers I made for legacy fluid items like `IFluidContainerItem` act on the ItemStack directly, and swap out the item with no problems. The new fluid util functions and documentation take this into account, and show that it works in a way that's not too ugly.

File: src/main/java/net/minecraftforge/fluids/capability/ItemFluidContainer.java (hunk 5 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/capability/ItemFluidContainer.java]: It's basically an extension of the documentation imo, and a nice win for capabilities. If you have the old `ItemFluidContainer` you see it's deprecated. The documentation there links to this new one. Oh wow, look how clean this is compared to the old one!

File: build.gradle (hunk 1 above)
Comment (mezz) [inline build.gradle]: Ah yeah I messed up my commit order, these extra things will be removed.

File: patches/minecraft/net/minecraft/item/ItemStack.java.patch (hunk 2 above)
Comment (mezz) [inline patches/minecraft/net/minecraft/item/ItemStack.java.patch]: Will be removed

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java (hunk 7 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: Will be removed, I had to update Forge's MCP to work with Forestry and so this slipped in.

**Comment by williewillus** [comment]
Ah in that case then a bump for merge from me
Teaching people about caps and IItemHandler always comes with a "is there a fluid version" so this would be great to have

File: src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java (hunk 4 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/ItemFluidContainer.java]: Added

**Comment by mezz** [comment]
Updated to address the first wave of comments.

File: src/main/java/net/minecraftforge/fluids/FluidUtil.java [hunk 8]
```diff
@@ -19,148 +37,453 @@ private FluidUtil()
...
+        boolean isDestNonSolid = !destMaterial.isSolid();
+        boolean isDestReplaceable = destBlockState.getBlock().isReplaceable(worldIn, pos);
+        if (!worldIn.isAirBlock(pos) && !isDestNonSolid && !isDestReplaceable)
+        {
+            return false; // Non-air, solid, unreplacable block. We can't put fluid here.
+        }
+
+        IBlockState fluidBlockState = fluid.getBlock().getDefaultState();
+
+        if (worldIn.provider.doesWaterVaporize() && fluidBlockState.getMaterial() == Material.WATER)
```
Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: I feel like this should have some kind of generic call back to the fluid/fluidblock object instead. For example, BC Oil explodes when placed in the nether.

File: src/main/java/net/minecraftforge/fluids/FluidUtil.java [hunk 9]
```diff
@@ -19,148 +37,453 @@ private FluidUtil()
...
-    public static boolean interactWithTank(ItemStack stack, EntityPlayer player, IFluidHandler tank, EnumFacing side)
+    /**
+     * Used to handle the common case of a fluid item right-clicking on a fluid handler.
+     * First it tries to fill the container item from the fluid handler,
+     * if that action fails then it tries to drain the container item into the fluid handler.
+     *
+     * Returns true if interaction was successful.
+     * Returns false if interaction failed.
+     */
+    public static boolean interactWithFluidHandler(ItemStack stack, IFluidHandler fluidHandler, EntityPlayer player)
```
Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: This doesn't seem to handle creative mode correctly, where you'd not want the container to be consumed.

File: src/main/java/net/minecraftforge/fluids/FluidUtil.java (hunk 8 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: I actually had that and removed it since I don't want to go of track with this PR. 
It will be easy to add separately. Just need a `fluid.doesVaporize()` and `fluid.vaporize(world, pos, blah)`.

File: src/main/java/net/minecraftforge/fluids/FluidUtil.java (hunk 9 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: Good catch, confirmed.

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: Fixed here https://github.com/MinecraftForge/MinecraftForge/pull/2923/commits/eb9b362312fc249208dc61ca11ba33b0f4dccd0c

**Comment by mezz** [comment]
Just noting there's still a bug with this PR that I'm working out:
I handled the `FluidContainerRegistry` containers item swapping, but not the case where the empty container is `null` aka "empty bucket" (wtf  4eb05ed5716085784c8ea0c29747f97d5b6ad2ae). Draining a container like this results in an empty bucket instead of destroying the item.

**Comment by mezz** [comment]
Fixed the bug with consumable fluid containers, they no longer leave a bucket in the player inventory.
No known bugs at this time.

**Comment by bonii-xx** [comment]
Looks like the changes killed the popup-animation when interacting with tanks.
From debugging this issue earlier in 1.9 where it was inconsistent with vanilla and forge buckets: This is because the client doesn't recognize the itemstack set from the packet sent by detectAndSendChanges inventory container as actually being changed because it already has the changed bucket set clientside. This packet takes care of the animation, but only if it's recognized as new and you can't do it clientside either too since the packet always replaces the itemstack, resetteng the itemstack.animationToGo variable.

File: src/main/java/net/minecraftforge/fluids/FluidUtil.java (hunk 8 above)
Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: Are all fluids still forced to be material water?

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidUtil.java]: No, I intend to replace the `fluidBlockState.getMaterial() == Material.WATER` with `fluid.doesVaporize()`.

**Comment by CovertJaguar** [comment]
So...if we plan to eventually delete FluidContainerRegistry entirely, how would IFluidContainerItem handle swapping ItemStacks? I see that @mezz has added support to the utility functions for nullable fluid containers, but FluidTankItemStack still seems to lack that ability. Meaning it is also still impossible to create a universal single use fluid container.

**Comment by mezz** [comment]
Thanks @bonii-xx I'll look into it. I didn't notice that the bucket was animating differently with a tank than when it picks up fluids.

@CovertJaguar FluidTankItemStack does not have that ability, but it's just a template.
I'll make another template for single-use fluid containers, something that I don't believe was possible before.  To clarify for anyone else (we talked on IRC), a single-use container can be filled, but when it's emptied it is destroyed. We've wanted that for Forestry's capsules but it wasn't really doable before.

**Comment by CovertJaguar** [comment]
While we are on the topic of fluids and containers, could we get an oreDict entry for empty buckets too? Since Forge now provides two versions.

Allowing Forge Ore Recipes to accept FluidStacks as a recipe ingredient would be awesome too. ;)
In this case the recipe parser would be looking for a fluid container with sufficient requested fluid in it.

**Comment by mezz** [comment]
You can't actually get the empty Forge Bucket now, it swaps to the vanilla bucket automatically. Same with lava and water buckets, there's only the vanilla kind.

**Comment by CovertJaguar** [comment]
Ah, ok cool. I had glanced over the code about a month or so ago but that wasn't immediately obvious.

**Comment by mezz** [comment]
For checking if a fluid container has enough of a fluid in it, maybe a helper like:
```java
// container must have stackSize == 1
public static boolean doesContainFluid(@Nonnull ItemStack container, @Nonnull FluidStack checkFluid) {
    IFluidHandler fluidHandler = FluidUtil.getFluidHandler(container);
    if (fluidHandler == null) {
        return false;
    }
    FluidStack containedFluid = fluidHandler.drain(fluid, false);
    return checkFluid.isFluidEqual(containedFluid);
}
```

**Comment by CovertJaguar** [comment]
I know I have similar code in my Fluid Utility class.

**Comment by CovertJaguar** [comment]
On the subject of the TankInfo method, one of its primary users is Buildcraft Gate Triggers. Or really anything that does some kind of non-transfer related inspection (computer mods?).

It is designed as a "safe" method to query the contents of a Tank without allowing the outside world to muck around with it directly.

I'd say it still has a place in the API.

**Comment by mezz** [comment]
The idea for tankInfo is to make something a bit more useful in its place. Right now the information is only slightly better than what you can get from simulating. 

It would be easy to add some more information about the tanks, like if they can input/output fluids, and what fluids they can hold. Something like BC gates, AE, OpenComputers, advanced fluid routing pipes, or info huds like TOP could potentially make use of that information. 

Forestry and others could also use the information to decide if a empty container can go in a certain slot. For example the Bottler doesn't want to accept fluid container items in its fill slot if they can't be filled. There is no way to correctly simulate to find that information because the Bottler can hold any fluid, there is no way to know for sure without simulating every known fluid.

**Comment by CovertJaguar** [comment]
More information is always good.

File: src/main/java/net/minecraftforge/fluids/FluidTank.java [hunk 10]
```diff
@@ -175,11 +198,38 @@ public FluidStack drain(int maxDrain, boolean doDrain)
...
         return stack;
     }
+
+    /**
+     * Returns true if the tank can be filled with this type of fluid.
+     * Used as a filter for fluid types.
+     * Does not consider the current state of the tank, only whether it could ever fill with this type of fluid.
+     */
+    public boolean canFillFluidType(FluidStack fluid)
+    {
```
Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: Why are these no longer part of IFluidHandler? Do you plan to replace them with the improved TankInfo stuff?

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: The original definition was unclear, many people (myself included) interpreted it as a quicker check like a simulated drain or fill, and it just opens the door for buggy fluid handling.

I want the API to be clearly divided between:

- "state" which is like an immutable overview of contents of the tank. 
- "properties" which would be properties that are not dependent on the state of the tank, and include transaction abilities like `canFillFluidType`,
-  "transaction" fill/drain, which lets you interact without needing to know state or info.

The current TankInfo mainly gives the "state" of the tank and should be expanded on.

Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: Sounds good.

File: src/main/java/net/minecraftforge/fluids/FluidTank.java [hunk 11]
```diff
@@ -1,14 +1,16 @@
...
+import javax.annotation.Nullable;
+
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 
 /**
  * Reference implementation of {@link IFluidTank}. Use/extend this or implement your own.
  */
-public class FluidTank implements IFluidTank
+public class FluidTank implements IFluidTank, net.minecraftforge.fluids.capability.IFluidHandler
```
Comment (liach) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: Why not use a normal import? This is not a patch.

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/FluidTank.java]: There is a non-capability `IFluidHandler` in the same package as FluidTank so this is just the editor default. I'll import it properly.

**Comment by mezz** [comment]
Updated the PR to add `IFluidTankProperties` which give a much better overview of what the fluid handlers contain and the basic rules they follow.

File: src/main/java/net/minecraftforge/fluids/capability/FluidTankProperties.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+	private final TankInteractionType interactionType;
+
+	public FluidTankProperties(@Nullable FluidStack contents, int capacity)
+	{
+		this(contents, capacity, TankInteractionType.OPEN);
+	}
+
+	public FluidTankProperties(@Nullable FluidStack contents, int capacity, TankInteractionType interactionType)
+	{
+		this.contents = contents;
```
Comment (CovertJaguar) [inline src/main/java/net/minecraftforge/fluids/capability/FluidTankProperties.java]: You'll want to make a copy of this rather than passing the original out. ;)

File: src/main/java/net/minecraftforge/fluids/capability/TankInteractionType.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+public enum TankInteractionType
+{
+	/** Open tanks can be filled or drained. */
+	OPEN,
+	/** Fill-only tanks can only be filled, not drained. */
+	FILL_ONLY,
+	/** Drain-only tanks can only be drained, not filled. */
+	DRAIN_ONLY,
+	/** Closed tanks can't be filled or drained. */
+	CLOSED
```
Comment (liach) [inline src/main/java/net/minecraftforge/fluids/capability/TankInteractionType.java]: Add boolean methods `canDrain` and `canFill`?

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/capability/TankInteractionType.java]: Yeah I agree. I was going to have more types but I ended up scaling it back to this to keep it simple. 
It would make more sense as booleans at this point.

Comment (gigaherz) [inline src/main/java/net/minecraftforge/fluids/capability/TankInteractionType.java]: Yep, as a general rule, if you are enumerating permutations, it may be that you are trying to mix two separate values into one.

Comment (mezz) [inline src/main/java/net/minecraftforge/fluids/capability/TankInteractionType.java]: This has been removed and replaced with boolean methods.

**Comment by mezz** [comment]
Updated here to point to the master branch: #2942

**Comment by Actuarius** [comment]
@mezz added labels [Fluids]
